### PR TITLE
Add LookupBlockIndex

### PR DESCRIPTION
**Pull request by promag**
Replace all `mapBlockIndex` lookups with the new `LookupBlockIndex()`. In some cases it avoids a second lookup.

### CODE REVIEW

File: src/rest.cpp [hunk 1]
```diff
@@ -212,10 +211,11 @@ static bool rest_block(HTTPRequest* req,
     CBlockIndex* pblockindex = NULL;
     {
         LOCK(cs_main);
-        if (mapBlockIndex.count(hash) == 0)
+        pblockindex = LookupBlockIndex(hash);
+        if (pblockindex) {
```
Comment (promag) [inline src/rest.cpp]: !pblockindex

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -669,10 +669,10 @@ UniValue getblockheader(const JSONRPCRequest& request)
     if (!request.params[1].isNull())
         fVerbose = request.params[1].get_bool();
 
-    if (mapBlockIndex.count(hash) == 0)
+    CBlockIndex* pblockindex = LookupBlockIndex(hash);
+    if (pblockindex == nullptr) {
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Same as above.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -817,7 +816,9 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
     stats.hashBlock = pcursor->GetBestBlock();
     {
         LOCK(cs_main);
-        stats.nHeight = mapBlockIndex.find(stats.hashBlock)->second->nHeight;
+        const CBlockIndex* pindex = LookupBlockIndex(stats.hashBlock);
+        assert(pindex);
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Remove assert.

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by TheBlueMatt** [comment]
This may make #10692 a bit simpler, though I'm not sure it matters what order they come in.

**Comment by promag** [comment]
Fixed bug and rebased.

**Comment by promag** [comment]
@laanwj I believe it's ready.

**Comment by laanwj** [comment]
Concept ACK

File: src/net_processing.cpp [hunk 4]
```diff
@@ -904,7 +906,7 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
         }
     case MSG_BLOCK:
     case MSG_WITNESS_BLOCK:
-        return mapBlockIndex.count(inv.hash);
+        return LookupBlockIndex(inv.hash);
```
Comment (MeshCollider) [inline src/net_processing.cpp]: Maybe explicit cast to bool here for clarity

**Review by meshcollider — COMMENTED** [review]
utACK

File: src/net_processing.cpp (hunk 4 above)
Comment (promag) [inline src/net_processing.cpp]: Prefer this?
```cpp
return LookupBlockIndex(inv.hash) != nullptr;
```

Comment (MeshCollider) [inline src/net_processing.cpp]: Yeah or a ternary or even an if statement would be a lot clearer

**Comment by promag** [comment]
Rebased and fixed @MeshCollider comment.

File: src/validation.h [hunk 5]
```diff
@@ -425,6 +425,12 @@ class CVerifyDB {
 /** Replay blocks that aren't fully applied to the database. */
 bool ReplayBlocks(const CChainParams& params, CCoinsView* view);
 
+inline CBlockIndex* LookupBlockIndex(const uint256& hash)
+{
+    BlockMap::const_iterator it = mapBlockIndex.find(hash);
```
Comment (promag) [inline src/validation.h]: Missing assert lock is held.

Comment (practicalswift) [reply src/validation.h]: Missing `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` too :-)

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -3793,7 +3791,8 @@ unsigned int CWallet::ComputeTimeSmart(const CWalletTx& wtx) const
 {
     unsigned int nTimeSmart = wtx.nTimeReceived;
     if (!wtx.hashUnset()) {
-        if (mapBlockIndex.count(wtx.hashBlock)) {
+        const CBlockIndex* pindex = LookupBlockIndex(wtx.hashBlock);
+        if (pindex) {
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp]: In some of these places you could do 

```c++
if (const CBlockIndex* pindex = LookupBlockIndex(wtx.hashBlock))
```

to reduce the scope of the pindex variables.

**Review by ryanofsky — COMMENTED** [review]
utACK 1d1768f2978bafaf166bfa218c899ddaeeccad2d

**Comment by promag** [comment]
Thanks @ryanofsky, I have to rebase and then I'll take into account your comment above.

File: src/init.cpp [hunk 7]
```diff
@@ -1434,8 +1434,9 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
 
                 // If the loaded chain has a wrong genesis, bail out immediately
                 // (we're likely using a testnet datadir, or the other way around).
-                if (!mapBlockIndex.empty() && mapBlockIndex.count(chainparams.GetConsensus().hashGenesisBlock) == 0)
```
Comment (jimpo) [inline src/init.cpp]: Need to keep the `!mapBlockIndex.empty()`.

File: src/net_processing.cpp [hunk 8]
```diff
@@ -2272,14 +2272,14 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
...
-        if (mapBlockIndex.find(cmpctblock.header.hashPrevBlock) == mapBlockIndex.end()) {
+        if (!LookupBlockIndex(cmpctblock.header.hashPrevBlock)) {
             // Doesn't connect (or is genesis), instead of DoSing in AcceptBlockHeader, request deeper headers
             if (!IsInitialBlockDownload())
                 connman->PushMessage(pfrom, msgMaker.Make(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexBestHeader), uint256()));
             return true;
         }
 
-        if (mapBlockIndex.find(cmpctblock.header.GetHash()) == mapBlockIndex.end()) {
+        if (LookupBlockIndex(cmpctblock.header.GetHash())) {
```
Comment (jimpo) [inline src/net_processing.cpp]: Needs a `!`

**Review by jimpo — COMMENTED** [review]
Concept ACK. This is much cleaner.

**Comment by promag** [comment]
This needs rebase. I can do it and adapt the new code if there is interest in merging it quick.

Maybe it should also make `BlockMap mapBlockIndex` static?

cc @MarcoFalke @laanwj

**Comment by TheBlueMatt** [comment]
Concept ACK. Would like to see this rebased sooner rather than later.

**Comment by promag** [comment]
Forgot to tackle @ryanofsky [comment](https://github.com/bitcoin/bitcoin/pull/11041#discussion_r148586107) above..

**Comment by MarcoFalke** [comment]
travis failure:

```
Assertion failed: lock cs_main not held in ./validation.h:433; locks held:

```

File: src/qt/transactionrecord.cpp [hunk 9]
```diff
@@ -167,10 +167,7 @@ void TransactionRecord::updateStatus(const CWalletTx &wtx)
     // Determine transaction status
 
     // Find the block the tx is in
-    CBlockIndex* pindex = nullptr;
-    BlockMap::iterator mi = mapBlockIndex.find(wtx.hashBlock);
-    if (mi != mapBlockIndex.end())
-        pindex = (*mi).second;
+    CBlockIndex* pindex = LookupBlockIndex(wtx.hashBlock);
```
Comment (laanwj) [inline src/qt/transactionrecord.cpp]: const?

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -674,10 +674,10 @@ UniValue getblockheader(const JSONRPCRequest& request)
     if (!request.params[1].isNull())
         fVerbose = request.params[1].get_bool();
 
-    if (mapBlockIndex.count(hash) == 0)
+    CBlockIndex* pblockindex = LookupBlockIndex(hash);
```
Comment (laanwj) [inline src/rpc/blockchain.cpp]: `const CBlockIndex*`?

File: src/rpc/blockchain.cpp [hunk 11]
```diff
@@ -753,12 +753,12 @@ UniValue getblock(const JSONRPCRequest& request)
             verbosity = request.params[1].get_bool() ? 1 : 0;
     }
 
-    if (mapBlockIndex.count(hash) == 0)
+    CBlockIndex* pblockindex = LookupBlockIndex(hash);
```
Comment (laanwj) [inline src/rpc/blockchain.cpp]: Another one that could be const. There are more, but I'll stop commenting every one.

**Review by laanwj — COMMENTED** [review]
- I think it would be preferable to use `const CBlockIndex*` where it's possible. 
- There are some `cs_main` locking changes; can you please explain them? Were locks missing there before, making this a bugfix as well as refactor?

**Comment by promag** [comment]
> I think it would be preferable to use const CBlockIndex* where it's possible

Agree.

> There are some cs_main locking changes; can you please explain them? Were locks missing there before, making this a bugfix as well as refactor?

Yes, for instance `FindForkInGlobalIndex`.

I'll check if adding annotations helps identifying more missing locks. I'll update accordingly.

**Comment by TheBlueMatt** [comment]
re: const CBlockIndex (and associated things): see #10692.

File: src/checkpoints.cpp [hunk 12]
```diff
@@ -21,9 +21,9 @@ namespace Checkpoints {
         for (const MapCheckpoints::value_type& i : reverse_iterate(checkpoints))
         {
             const uint256& hash = i.second;
-            BlockMap::const_iterator t = mapBlockIndex.find(hash);
-            if (t != mapBlockIndex.end())
-                return t->second;
+            if (CBlockIndex* const pindex = LookupBlockIndex(hash)) {
```
Comment (TheBlueMatt) [inline src/checkpoints.cpp]: I dont think that is the type of const you meant. Probably just leave it non-const for now and I'll fix it up in #10692 (or you can start pulling stuff out of there into this PR if you prefer, but that probably just ends up taking more time).

File: src/init.cpp [hunk 13]
```diff
@@ -1487,6 +1492,8 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
 
                 bool is_coinsview_empty = fReset || fReindexChainState || pcoinsTip->GetBestBlock().IsNull();
                 if (!is_coinsview_empty) {
+                    LOCK(cs_main);
```
Comment (TheBlueMatt) [inline src/init.cpp]: This seems unrelated. Maybe leave it for #11596 instead? Or just do as I suggested at https://github.com/bitcoin/bitcoin/pull/11596#discussion_r155811677 and put a cs_main at the top of the load-chain-data block here.

File: src/validation.cpp [hunk 14]
```diff
@@ -259,12 +259,12 @@ namespace {
 
 CBlockIndex* FindForkInGlobalIndex(const CChain& chain, const CBlockLocator& locator)
 {
+    LOCK(cs_main);
```
Comment (TheBlueMatt) [inline src/validation.cpp]: Looks like you're just adding this for CreateWalletFromFile, which already needs cs_main (and is missing it there). This should be an AssertLockHeld.

File: src/validation.cpp [hunk 15]
```diff
@@ -1813,9 +1813,9 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
         //  relative to a piece of software is an objective fact these defaults can be easily reviewed.
         // This setting doesn't force the selection of any particular chain but makes validating some faster by
         //  effectively caching the result of part of the verification.
-        BlockMap::const_iterator  it = mapBlockIndex.find(hashAssumeValid);
-        if (it != mapBlockIndex.end()) {
-            if (it->second->GetAncestor(pindex->nHeight) == pindex &&
+        const CBlockIndex* pindex_valid = LookupBlockIndex(hashAssumeValid);
```
Comment (TheBlueMatt) [inline src/validation.cpp]: Dunno if its worth bothering for things in CChainState. The whole goal here is to make mapBlockIndex a private in CChainState and expose LookupBlockIndex externally.

File: src/validation.cpp [hunk 16]
```diff
@@ -3641,17 +3640,20 @@ CBlockIndex * CChainState::InsertBlockIndex(const uint256& hash)
     if (hash.IsNull())
         return nullptr;
 
+    LOCK(cs_main);
```
Comment (TheBlueMatt) [inline src/validation.cpp]: This should absolutely be an AssertLockHeld, if anything. If you add the LOCK(cs_main) in the block-index-loading code in init.cpp this should be able to go away.

File: src/validation.cpp [hunk 17]
```diff
@@ -4230,24 +4236,31 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, CDiskB
...
-                    LogPrint(BCLog::REINDEX, "%s: Out of order block %s, parent %s not known\n", __func__, hash.ToString(),
-                            block.hashPrevBlock.ToString());
-                    if (dbp)
-                        mapBlocksUnknownParent.insert(std::make_pair(block.hashPrevBlock, *dbp));
-                    continue;
-                }
 
-                // process in case the block isn't known yet
-                if (mapBlockIndex.count(hash) == 0 || (mapBlockIndex[hash]->nStatus & BLOCK_HAVE_DATA) == 0) {
+                {
```
Comment (TheBlueMatt) [inline src/validation.cpp]: Same here. Can go away with an init.cpp LOCK(cs_main).

**Review by TheBlueMatt — COMMENTED** [review]
Dont know how much its worth focusing on making all the CBlockIndex*s const until we make LookupBlockIndex return a const pointer, which requires a few additional commits that will just delay this PR. For the ones that can be done trivially it'd be nice, but I wouldn't say its worth delaying this PR for it as they'll just be scripted-diff'd in #10692.

File: src/validation.cpp (hunk 16 above)
Comment (promag) [reply src/validation.cpp]: Do we want to lock that long?

File: src/validation.cpp (hunk 14 above)
Comment (promag) [reply src/validation.cpp]: And why does `CreateWalletFromFile` needs `cs_main`?

Comment (TheBlueMatt) [reply src/validation.cpp]: Hmm, there was some discussion of this in some other PR recently, maybe a @practicalswift PR? In any case, there are a number of things in CreateWalletFromFile that need cs_main (though possibly we want to avoid it and just take it for parts to avoid cs_main during the rescan after #11281), eg chainActive.GetLocator(), chainActive.Genesis (maybe not technically, but probably), FindForkInGlobalIndex, chainActive.Tip(), etc. So definitely worth just taking cs_main there and making this an AssertLockHeld.

File: src/validation.cpp (hunk 16 above)
Comment (TheBlueMatt) [reply src/validation.cpp]: It doesnt matter, its early in init long before anything else is going to take cs_main, so taking it should have no impact.

File: src/init.cpp (hunk 13 above)
Comment (promag) [reply src/init.cpp]: It's needed for the assert below (L1502).

File: src/validation.cpp (hunk 15 above)
Comment (promag) [reply src/validation.cpp]: I'd prefer to avoid special cases for now. Eventually `LookupBlockIndex` will loose `inline`, so do it then?

File: src/validation.cpp (hunk 17 above)
Comment (promag) [reply src/validation.cpp]: I don't think so, this can be called from RPC. Also, below there is  `ReadBlockFromDisk` and we want to avoid cs_main lock there right?

File: src/init.cpp (hunk 13 above)
Comment (TheBlueMatt) [reply src/init.cpp]: Looks like you can remove this line now?

File: src/validation.cpp [hunk 18]
```diff
@@ -2762,11 +2762,14 @@ bool ResetBlockFailureFlags(CBlockIndex *pindex) {
...
+    AssertLockHeld(cs_main);
+
     // Check for duplicate
     uint256 hash = block.GetHash();
-    BlockMap::iterator it = mapBlockIndex.find(hash);
-    if (it != mapBlockIndex.end())
-        return it->second;
+    CBlockIndex* pindex = LookupBlockIndex(hash);
+    if (pindex) {
+      return pindex;
```
Comment (TheBlueMatt) [inline src/validation.cpp]: nit: bad spacing here.

File: src/validation.cpp (hunk 15 above)
Comment (TheBlueMatt) [reply src/validation.cpp]: For cases where we are fine with a const CBlockIndex* result, I dont think it matters much, but I'd kinda like to have LookupBlockIndex only return const CBlockIndex*es, so for things like CChainState::AddToBlockIndex, we'd have to partially revert this change then, easier to not do it in the first place IMO.

File: src/wallet/wallet.cpp [hunk 19]
```diff
@@ -3960,6 +3963,8 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
     // Try to top up keypool. No-op if the wallet is locked.
     walletInstance->TopUpKeyPool();
 
+    LOCK(cs_main);
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: Hmm, this is gonna conflict partially with #11281. The neccessary refactors seem somewhat unrelated, so I'd be fine with pushing the conflict resolution onto #11281, but if #11281 lands first this is gonna need a bit of tweaking...up to you if you want to do it now.

File: src/validation.cpp (hunk 15 above)
Comment (promag) [reply src/validation.cpp]: > have LookupBlockIndex only return const CBlockIndex*es

If you look closely you will see that a lot of function signatures require non-const `CBlockIndex*`, although a const should do the work (almost?) everywhere. I'm happy to do that in a follow up and deal with the non-const cases there. At least this PR is *complete*.

**Comment by markaw67** [comment]
THANK YOU!!!!

On Jan 2, 2018 8:14 PM, João Barbosa <notifications@github.com> wrote:

@promag commented on this pull request.

________________________________

In src/validation.cpp<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbitcoin%2Fbitcoin%2Fpull%2F11041%23discussion_r159353223&data=02%7C01%7C%7C07969f713b7c4a64bf1c08d5524fc83b%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636505424966541060&sdata=uFeOBfvlIXBm%2FB6rVQAizqsjsaBYjNjTFasd0O1t9fM%3D&reserved=0>:

> @@ -1813,9 +1813,9 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
         //  relative to a piece of software is an objective fact these defaults can be easily reviewed.
         // This setting doesn't force the selection of any particular chain but makes validating some faster by
         //  effectively caching the result of part of the verification.
-        BlockMap::const_iterator  it = mapBlockIndex.find(hashAssumeValid);
-        if (it != mapBlockIndex.end()) {
-            if (it->second->GetAncestor(pindex->nHeight) == pindex &&
+        const CBlockIndex* pindex_valid = LookupBlockIndex(hashAssumeValid);


have LookupBlockIndex only return const CBlockIndex*es

If you look closely you will see that a lot of function signatures require non-const CBlockIndex*, although a const should do the work (almost?) everywhere. I'm happy to do that in a follow up and deal with the non-const cases there. At least this PR is complete.

**Review by ryanofsky — COMMENTED** [review]
utACK 77cc5b7f3dd1c663d6a16946563c7b59d0b3184a. Main change since last review is locking cs_main in AppInitMain, adding lock asserts.

File: src/init.cpp [hunk 20]
```diff
@@ -1408,6 +1408,8 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
 
         uiInterface.InitMessage(_("Loading block index..."));
 
+        LOCK(cs_main);
```
Comment (TheBlueMatt) [inline src/init.cpp]: This should be in a separate commit.

File: src/validation.cpp [hunk 21]
```diff
@@ -260,12 +260,12 @@ namespace {
 
 CBlockIndex* FindForkInGlobalIndex(const CChain& chain, const CBlockLocator& locator)
 {
+    AssertLockHeld(cs_main);
```
Comment (TheBlueMatt) [inline src/validation.cpp]: This could go in a separate commit.

File: src/validation.cpp (hunk 15 above)
Comment (TheBlueMatt) [reply src/validation.cpp]: Making everything const isnt so hard, see #10692, which did it mostly in scripted-diffs.

File: src/validation.cpp [hunk 22]
```diff
@@ -2772,11 +2772,14 @@ bool ResetBlockFailureFlags(CBlockIndex *pindex) {
...
 CBlockIndex* CChainState::AddToBlockIndex(const CBlockHeader& block)
 {
+    AssertLockHeld(cs_main);
+
     // Check for duplicate
     uint256 hash = block.GetHash();
-    BlockMap::iterator it = mapBlockIndex.find(hash);
-    if (it != mapBlockIndex.end())
-        return it->second;
+    CBlockIndex* pindex = LookupBlockIndex(hash);
```
Comment (TheBlueMatt) [inline src/validation.cpp]: Again, please do not use LookupBlockIndex here (and a few other places). If the next step is to make LookupBlockIndex return a const CBlockIndex* (which I think it clearly is), this part of the patch just has to get reverted. There is no reason to avoid using mapBlockIndex in CChainState members, that's what CChainState is there for :).

**Review by TheBlueMatt — COMMENTED** [review]
Please make locking changes at least be a separate commit. Please don't use LookupBlockIndex when the CBlockIndex can't be made const ala #10692.

File: src/validation.cpp (hunk 22 above)
Comment (promag) [reply src/validation.cpp]: > There is no reason to avoid using mapBlockIndex in CChainState members, that's what CChainState is there for

At that point there can be `CBlockIndex* CChainState::LookupBlockIndex()`?

File: src/init.cpp (hunk 20 above)
Comment (promag) [reply src/init.cpp]: Done.

File: src/validation.cpp (hunk 21 above)
Comment (promag) [reply src/validation.cpp]: Done.

File: src/validation.cpp (hunk 22 above)
Comment (TheBlueMatt) [reply src/validation.cpp]: Why? I see no reason to have a LookupBlockIndex at all *except* to avoid exposing mapBlockIndex/non-const CBlockIndexes outside of CChainState/validation.cpp. If we're inside of CChainState, doing a LookupBlockIndex only serves to add needless indirection.

**Comment by promag** [comment]
> Again, please do not use LookupBlockIndex here (and a few other places).

As requested, last commit reverts changes to `CChainState`. Please review.

**Review by ryanofsky — COMMENTED** [review]
utACK 7326660786b71af254c2c4bc40c7563be060db8a. Changes up to the second to last commit (d889c036cd...18144408b6) are identical to what I previously reviewed modulo a rebase, commit splitting, and reverting a whitespace change. Very last commit (7326660786b71af254c2c4bc40c7563be060db8a) reverts some changes in the second to last commit, as requested by Matt. Looks fine.

**Comment by promag** [comment]
Thanks @ryanofsky.

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (kallewoof) [reply src/wallet/wallet.cpp]: I didn't know you could declare inside an if statement. Cool.

**Review by kallewoof — COMMENTED** [review]
utACK -- Nice clean up!

**Comment by promag** [comment]
Rebased.

@kallewoof why drop `const` qualifier?

**Comment by kallewoof** [comment]
@promag Sorry, I only meant "move the `*`"

**Comment by promag** [comment]
@kallewoof no problem, done.

**Comment by kallewoof** [comment]
utACK 92fabcd

**Comment by ryanofsky** [comment]
utACK 92fabcd443322dcfdf2b3477515fae79e8647d86. No changes since last review except rebase and squash.

File: src/wallet/wallet.cpp [hunk 23]
```diff
@@ -3816,7 +3814,12 @@ unsigned int CWallet::ComputeTimeSmart(const CWalletTx& wtx) const
 {
     unsigned int nTimeSmart = wtx.nTimeReceived;
     if (!wtx.hashUnset()) {
-        if (mapBlockIndex.count(wtx.hashBlock)) {
+        const CBlockIndex* pindex = nullptr;
+        {
+            LOCK(cs_main);
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp]: I think this new lock is causing test failures on master. If I run

```
./configure --enable-debug
make -j12 -C src test/test_bitcoin && src/test/test_bitcoin --log_level=test_suite --run_test=wallet_tests/ComputeTimeSmart
```

I see:

```
Entering test case "ComputeTimeSmart"
test_bitcoin: sync.cpp:100: void potential_deadlock_detected(const std::pair<void*, void*>&, const LockStack&, const LockStack&): Assertion `false' failed.
unknown location(0): fatal error in "ComputeTimeSmart": signal: SIGABRT (application abort requested)
wallet/test/wallet_tests.cpp(566): last checkpoint
```

The problem is that ComputeTimeSmart is called by CWallet::AddToWallet, which only acquires cs_wallet, so cs_main will wind up being acquired after cs_wallet during the test.

**Comment by rebroad** [comment]
@promag What's the reason for avoiding mapBlockIndex lookups?

**Comment by promag** [comment]
@rebroad it avoids duplicate lookups - its unnecessary work.
